numpy/_core/src/multiarray/stringdtype/sorts.c

```diff
@@ -0,0 +1,321 @@
+    if (elsize == 0) {
+        return 0;
+    }
+
+    pl = (npy_intp *)data[1];
+    pr = pl + num;
+    pw = (npy_intp *)malloc((num >> 1) * sizeof(npy_intp));
+    if (pw == NULL) {
+        return -2;
+    }
+    string_amergesort0(pl, pr, (char *)data[0], pw, elsize, sdescr);
+    free(pw);
+    NpyString_release_allocator(allocator);
+
+    return 0;
+}
+
+int
+stringdtype_argsort_get_loop(PyArrayMethod_Context *context, int aligned,
+                             int move_references, const npy_intp *strides,
+                             PyArrayMethod_StridedLoop **out_loop,
+                             NpyAuxData **out_transferdata,
+                             NPY_ARRAYMETHOD_FLAGS *flags)
+{
+    PyArrayMethod_SortParameters *parameters =
+            (PyArrayMethod_SortParameters *)context->parameters;
+
+    *flags |= NPY_METH_NO_FLOATINGPOINT_ERRORS;
+
+    switch (parameters->flags) {
+        case NPY_SORT_DEFAULT:
+            *out_loop = &stringdtype_stableargsort;  // default to mergesort
+            break;
+        case NPY_SORT_STABLE:
+            *out_loop = &stringdtype_stableargsort;
+            break;
+        default:
+            npy_gil_error(PyExc_ValueError, "Unsupported sort kind for StringDType");
+            return -1;
+    }
+
+    return 0;
+}
+
+int
+init_stringdtype_sorts(void)
+{
+    PyArray_DTypeMeta *dtypes[2] = {&PyArray_StringDType, &PyArray_StringDType};
+    PyType_Slot sort_slots[2] = {{NPY_METH_get_loop, &stringdtype_sort_get_loop},
+                                 {0, NULL}};
+    PyArrayMethod_Spec sort_spec = {
+            .name = "stringdtype_sort",
+            .nin = 1,
+            .nout = 1,
+            .dtypes = dtypes,
+            .slots = sort_slots,
+            .casting = NPY_NO_CASTING,
+            .flags = NPY_METH_NO_FLOATINGPOINT_ERRORS,
+    };
+
+    PyType_Slot argsort_slots[2] = {{NPY_METH_get_loop, &stringdtype_argsort_get_loop},
+                                    {0, NULL}};
+    PyArrayMethod_Spec argsort_spec = {
+            .name = "stringdtype_argsort",
+            .nin = 1,
+            .nout = 1,
+            .dtypes = dtypes,
+            .slots = argsort_slots,
+            .casting = NPY_NO_CASTING,
+            .flags = NPY_METH_NO_FLOATINGPOINT_ERRORS,
+    };
+
+    PyBoundArrayMethodObject *sort_method = PyArrayMethod_FromSpec_int(&sort_spec, 1);
+    if (sort_method == NULL) {
+        return -1;
+    }
+    NPY_DT_SLOTS(&PyArray_StringDType)->sort_meth =
+            (PyArrayMethodObject *)sort_method->method;
+    Py_INCREF(sort_method->method);
+    Py_DECREF(sort_method);
+
+    PyBoundArrayMethodObject *argsort_method = PyArrayMethod_FromSpec_int(&argsort_spec, 1);
+    if (argsort_method == NULL) {
+        return -1;
+    }
+    NPY_DT_SLOTS(&PyArray_StringDType)->argsort_meth =
+            (PyArrayMethodObject *)argsort_method->method;
+    Py_INCREF(argsort_method->method);
+    Py_DECREF(argsort_method);
+
+    return 0;
```

Comment:
Unfortunately we cannot use `PyUFunc_AddLoopsWithSpecs` if we group this with the ufuncs because NumPy is not fully initialized when multiarray is initialized.

Reply:
see https://github.com/numpy/numpy/pull/30112/files#r2501032204